src/main/java/io/axoniq/axonserver/connector/event/impl/BufferedPersistentStreamSegment.java

```diff
@@ -0,0 +1,128 @@
+package io.axoniq.axonserver.connector.event.impl;
+
+import io.axoniq.axonserver.connector.event.PersistentStreamSegment;
+import io.axoniq.axonserver.connector.impl.StreamClosedException;
+import io.axoniq.axonserver.grpc.event.EventWithToken;
+
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.LongConsumer;
+
+public class BufferedPersistentStreamSegment
+        implements PersistentStreamSegment {
+
+    private static final Runnable NO_OP = () -> {
+    };
+
+    private final Set<Runnable> onSegmentClosedCallbacks = new CopyOnWriteArraySet<>();
+    private final BlockingQueue<EventWithToken> buffer = new LinkedBlockingQueue<>();
+    private final AtomicReference<Throwable> errorResult = new AtomicReference<>();
+    private final AtomicBoolean closed = new AtomicBoolean();
+
+    private final AtomicReference<Runnable> onAvailableCallback = new AtomicReference<>(NO_OP);
+
+    private final int segment;
+    private final LongConsumer progressCallback;
+
+    public BufferedPersistentStreamSegment(int segment, LongConsumer progressCallback) {
+        this.segment = segment;
+        this.progressCallback = progressCallback;
+    }
+
+    @Override
+    public EventWithToken peek() {
+        checkClosed();
+        return buffer.peek();
+    }
+
+    @Override
+    public EventWithToken nextIfAvailable() {
+        checkClosed();
```

Comment:
I wonder if instead of checking the value of an AtomicInteger each time we access the queue, we could achieve better performance by using a 'special case' placeholder that represents the 'stream closed' situation. When that element is taken off the stream, it should be placed back and we should check the error fields to see how the stream was closed.